README.rst
======
orvara
======
This repo contains orvara, the package for fitting orbits of bright stars and
their faint companions (exoplanets, brown dwarfs, white dwarfs, and low-mass
stars).


Installation
============
NOTE: you must have Cython installed prior to installing orvara. The following
instructions will git clone the repo, then install Cython, and then install orvara:

.. code-block:: bash

    git clone https://github.com/t-brandt/orvara
    cd orvara
    pip install cython
    pip install -e .


Verifying
---------
cd to the root directory of the repo (if you are not already there). Run:

.. code-block:: bash

    pytest -sv

This will run a small suite of tests. This should take about 5 minutes. If any
of the tests fail, something is wrong with the install. Maybe you recently
switched branches? Try deleting the two compiled c files (orbit.c and orbit.cpython-38...)
and then run ``pip install -e .`` again. If the issue
persists, please submit an issue ticket on github!

Orbit fitting
=============

Configuration
-------------
First, assign the appropriate file directories and settings inside of a
config.ini file. Example configuration files can be found in
``orvara/tests/`` e.g. ``orvara/tests/config.ini``.

Epoch Astrometry
~~~~~~~~~~~~~~~~
If you are using epoch astrometry (observational epochs and scan angles), you
must give paths for ``GaiaDataDir``, ``Hip1DataDir``, and
``Hip2DataDir``. Those are the paths of the intermediate epoch and scan
angles for Gaia/GOST in csv format, the intermediate astrometry from the
original Hipparcos data reduction, and from the second Hipparcos data
reduction.

We recommend starting with ``use_epoch_astrometry = False``. If this
fails, then there is something wrong with the RVFile, HGCAFile, or (relative)
AstrometryFile. If that chain finishes fine, then set ``use_epoch_astrometry = True``.


It is preferred if your data come from the Java Tool (published in 2021). You will want to point orvara to the directory
that contains the H00, H01, etc. folders. It is also OK to have all the IAD in one folder, in which case, point orvara
to that large folder.
For more information about the Java Tool IAD, see:
https://www.cosmos.esa.int/web/hipparcos/interactive-data-access and https://www.cosmos.esa.int/web/hipparcos/catalogues
you can download the data there as a zip file. See https://arxiv.org/abs/2109.06761 Brandt et al. 2021 for
the explanation as to why the Java Tool data are preferred. We recommend only using the Java Tool IAD. If you only use
the Java Tool IAD, then ignore the following three warnings.

DVD IAD Warning 1: If your Hip2 intermediate data come from the DVD, you will
want to point to the 'resrec' folder. This should be e.g.:
``Hip2_DVD_Book/IntermediateData/resrec``. You can keep the original
directory structure as it is.

DVD IAD Warning 2: If you have fits of both the Java Tool and DVD IAD, then
it is imperative that you keep those different IAD in different directories. The DVD IAD and Java tool IAD have
slightly different data formats. Orvara will abort if you have more than one IAD
file with the same hipparcos id. (E.g., a folder with both H003850.d and HIP003850.d)

DVD IAD Warning 3: As described in https://arxiv.org/abs/2109.06761 Brandt et al. 2021,
there are roughly 6600 sources whose DVD IAD is suspected to be corrupted. This corruption is automatically
fixed by htof if you use the Java Tool IAD.

Setting priors
~~~~~~~~~~~~~~
Adding Gaussian mass priors are supported. As well, bounded log-uniform priors
on the RV jitter are supported. To add a Gaussian mass prior of 1 ± 0.1 M_sun
on the primary, you will want to add the following section to your
configuration file:

.. code-block:: bash

    [priors_settings]
    mpri = 1
    mpri_sig = 0.1
    minjitter = 1e-5
    maxjitter = 1e3

``minjitter`` and ``maxjitter`` are the lower and upper bounds respectively for
the log-uniform prior. E.g. If you wanted to set a log-uniform prior on jitter
between 1 and 300 meters/second (and zero outside that prior), you would set
``minjitter = 1`` and ``maxjitter = 300``.

``mpri`` is the mean of the gaussian prior on the primary mass. ``mpri_sig`` is
the standard deviation of that distribution. E.g. for our example of a gaussian
prior on the primary mass of 1 solar mass and 0.1 solar mass deviation, you set
``mpri = 1`` and ``mpri_sig = 0.1``.

Leaving ``mpri_sig = inf`` will turn off the prior and default to a 1/m prior.

Fits without using HGCA (hipparcos and gaia) absolute astrometry
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can fit an object that does not exist in the HGCA. Simply comment out the ``HipID`` line
in the configuration file. E.g.,:

.. code-block:: text

    . more stuff above...
    .
    # In that case you must supply a parallax and parallax_error in [priors_settings].
    #HipID = 3850
    #^ # note that we have commented the hip id out!!!
    .
    . more stuff below...


And set a parallax and parallax error in the priors section. So:

.. code-block:: text

    . more stuff above...
    .
    [priors_settings]

    parallax=53.05263
    parallax_error=2.818e-02
    .
    . more stuff below...

And that is it! The fit will proceed as normal, but gaia and hipparcos
epoch astrometry *will not* constrain the orbit.


Fits without subsets of data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can do fits with orvara to any subset of the three data sources:
1. Radial velocities
2. Absolute astometry (via the HGCA)
3. Relative astrometry.

E.g., you can do fits without the absolute astrometry (so just 1. and 3.), or a fit to purely radial velocities. Etc.
Below are descriptions with how to do fits to a source without a specific set of data.

1. To do a fit without radial velocity data: In the config file, set ``RVFile = None``
2. To do a fit without Absolute astrometry: In the config file, set ``HipID = None``
3. To do a fit without Relative Astrometry: ``AstrometryFile = None``

So if I wanted to do a fit to just radial velocities, I would set ``HipID = None`` and ``AstrometryFile = None``, and
``RVFile = path/to/the_rv_data.dat``.

Starting conditions
~~~~~~~~~~~~~~~~~~~
You can set the initial conditions of the orbit (starting parameters of the
walkers of the MCMC chain) via a starting file specified in the config.ini
file. There is an example starting file provided in ``orvara/tests/start_Gl758.dat``.

Usage
-----
After setting paths and MCMC (markov-chain monte-carlo)  settings in a
config.ini file, you fit an orbit by running the following from the command
line.

.. code-block:: bash

    fit_orbit --output-dir /path/to/output path/to/config.ini

If you do not specify an output directory using ``--output-dir``, then
orvara will write its output files to the current working directory. One can
set the number of threads in the config.ini file via ``nthreads``. Note
that the built-in parallelization is poor, and that parallelization might not
work at all on some systems. If you get an error when running the code check to
see if it goes away when setting ``nthreads=1``. It is often better to set
``nthreads`` to ``1`` then simply run multiple instances of orvara on separate
cores.


You can access the help menu with the --help flag as follows.

.. code-block:: bash

    fit_orbit --help


Input Data
~~~~~~~~~~

Orvara can fit orbits to a combination of input data. Accepted are (any combination of): relative radial velocities
(the radial velocity of the planet minus the radial velocity of the star), radial velocities, relative astrometry.

See orvara/tests/data for examples of these input data files. Be sure that the units are consistent
with those presented in the examples.

Orvara will automatically include absolute astrometric constraints when available. If the star is not inside of the
Hipparcos Gaia Catalog of Accelerations, these data will be excluded in the fit.

Output
~~~~~~
The output of the MCMC is a .fits file contained within your given output
directory (default current working directory). The name of the file will be
given by the ``target`` field within the ``[plotting]`` field of the
config file, with ``_chain000.fits`` appended (the integer will be
incremented to the next free filename). If this field is empty, the backup
target name will follow the form ``HIP%d_chain%03d.fits``.

The output file contains two .fits extensions: an empty one, and a fits table
with all the MCMC parameters at each step of the chain, together with the log
of the (unnormalized) posterior probability and some additional quantities.

HDU0: empty
"""""""""""
The first extension is empty for table data.

HDU1: table
"""""""""""
This is a fits table object. Each table column is of shape ``(nwalkers,
nsteps/thin)`` where ``thin`` is the thinning used in the configuration file (default
50, to save every 50th step). You may access a column by, e.g.,

``lnlike = hdulist[1].data['lnp']``

The column names and descriptions are:

    'mpri' : Primary mass (Solar masses)

    'msec0' : Secondary mass of the first (index 0) companion, Solar masses

    'sau0' : Semimajor axis of the first companion, AU

    'esino0' : sqrt(ecc)*sin(omega) for the first companion

    'ecoso0' : sqrt(ecc)*cos(omega) for the first companion

    'inc0' : inclination (radians) for the first companion

    'asc0' : PA of the ascending node (radians) for the first companion

    'lam0' : Mean longitude at reference epoch for the first companion

If there is more than one companion, then there are additional fields with,
e.g., 'msec1', 'msec2', etc.

    'jitter' : log RV jitter in m/s

    'jitter0' : log RV jitter for instrument 0

Note that 'jitter0', 'jitter1', etc. are present and 'jitter' is not if using
one jitter per instrument. The default is to use the same jitter for all
instruments. In this case 'jitter' is present but 'jitter0', 'jitter1', etc.
are not.

    'lnp' : natural log of the (unnormalized) probability. Note that this includes matrix determinants and is not simply chi squared.

    'plx_ML' : maximum likelihood (ML) parallax at this chain step

    'pmra_ML' : ML proper motion in RA at this chain step

    'pmdec_ML' : ML proper motion in Dec at this chain step

    'chisq_sep' : The chi squared in separation at the ML parallax at this chain step

    'chisq_PA' : The chi squared in position angle at this chain step

    'chisq_H' : The chi squared for the two Hipparcos proper motions

    'chisq_HG' : The chi squared for the two long-term Hipparcos-Gaia proper motions

    'chisq_G' : The chi squared for the two Gaia proper motions

    'chisq_relRV' : The chi squared for the relative RV data (RV planet - RV star). This will be zero if no relative RV data is used. It will be "None" if the chain you are plotting is from an older version of orvara which did not have relRV data support.

    'RV_ZP_0_ML' : The ML zero point (barycenter RV) for instrument 0

There will be an 'RV_ZP_1_ML' for instrument 1, etc., up to the number of RV instruments.

NOTE: chisq_RV, the chisquared of the fit to the RV data *is not saved*. This is for the following reason. Orvara
uses a jitter term such that the reduced chisquared ( equal to (chisquared of RV fit)/(number of RV data points) ) is 1. So
if you want chisq_RV, it is simply a value very close to the number of RV data points.

If you want an overall absolute astrometric chi squared, you would add the
values from items 'chisq_H', 'chisq_HG', and 'chisq_G' above. There are
effectively four measurements since the mean proper motion of the system was
fit ('pmra_ML' and 'pmdec_ML').

For instance, displaying ``hdulist[1].data['plx_ML']`` will show all the
walkers for the parallax chain (however this parameter is marginalized over in
orvara, it is not fit). ``numpy.mean(hdulist[1].data['plx_ML'][:, burn:])``
and ``numpy.std(hdulist[1].data['plx_ML'][:, burn:])`` would give the mean
and standard deviation of the maximum likelihood parallax (with burn = some
integer that is the number of steps/thinning factor that you are discarding as
burn in). You should add the measurement error of the parallax in quadrature
with this particular uncertainty.

One can use the 'lnp' column to compare the likelihoods of the best orbits if a
certain posterior is multimodal. Assume that the marginalized posterior in PA
is multimodal, with a mode at a value > 180 degrees, and a mode at a value less
than 180 degrees. The following code would print the likelihoods of the maximum
likelihood orbits at each of those two modes.


.. code-block:: python

    tt = fits.open('my/path/chain.fits')[1].data
    logl = tt['lnp']
    pa_data = (tt['asc0']*180/np.pi) % 360
    ls180 = pa_data < 180
    g180 = pa_data >= 180

    print(np.max(logl[ls180]))  # the max log likelihood of all orbits with PA of ascending nodes < 180 degrees
    print(np.max(logl[g180]))  # the max log likelihood of all orbits with PA of ascending nodes > 180 degrees


Examples
--------
To run a quick test using the test data and test config.ini in orvara/tests,
you could cd to the root directory of orvara, then run the following

.. code-block:: bash

    fit_orbit orvara/tests/config.ini --output-dir ~/Downloads

This will create a .fits file in the Downloads folder. The MCMC should
terminate in less than one second because of the short number of steps
indicated in the example config file.

The end-to-end tests in test_e2e check that the code is converging to
previously accepted values for HIP3850. If you wanted to run the code yourself
on this test case and check the results yourself against those in
misc/Diagnostic_plots.ipynb, you can run:

.. code-block:: bash


CHANGES.md
1.1.4 (2022-08-23)
------------------
- Changed the internals of the astrometric prediction code to use the orvara internals. This is 
more robust and works for more than just 2-body systems.
- A position (ra, dec) and uncertainty is now printed to standard out with the astrometric plot. 

1.1.3 (2022-05-24)
------------------
- Added the ability to constrain orbits with a relative RV measurement (e.g. that between beta pic A and beta
pic b), by providing a relative RV file (`relRVFile=relative_rv_filename.dat`) in the config.ini. This is implemented
programmatically so that it will work for multiple companions (3, 4 etc body fits), i.e., it won't crash.
However the logic to compute the relative RV is technically only valid in detail for 2-body systems, or multi-planet
systems where the primary is substantially more massive than any companion. See the note within `def calc_relRV` inside
of orbit.pyx.
- the chisquared of the relative RV fit is now saved (in the same way as the hipparcos/gaia chisquareds).

1.1.2 (2022-04-28)
------------------
- Bugfix for plotting proper motions. When the internals changed to mas/yr with 1.1.1, the plotting 
was still done in units of mas/day.

1.1.1 (2022-04-21)
------------------
- Added support for the Java Tool IAD (colloquially called hip21).

1.1.0 (2022-04-21)
------------------
- Added Seven and nine parameter fits in preparation for Gaia DR3, with tests in test_main.py.
- Added informative comments to the log likelihood function.
- Changed the astrometric fitting routine to work natively in units of years, instead of days. This
improved the condition number of the matrix that is solved via SVD.
- Bumped required version of HTOF to 1.1.0. This is needed for the fits with parallax.

1.0.5 (2021-12-06)
------------------
- Fixed a bug in the astrometric orbit plotting.

1.0.4 (2021-05-14)
------------------
- Fixed units bug in periastron time calculation.

1.0.3 (2021-05-14)
------------------
- Another small tweak to secondary priors: ensure 1/M if a given prior is not set.

1.0.2 (2021-05-13)
------------------
- Fixed a memory leak when fitting multiple companions, fixed prior on secondary masses to fully undo a 1/M prior.

1.0.1 (2021-04-28)
------------------
- Fixed the diagnostic test in test_main.py so that it will work on windows. File i/o failed on the
test due to https://github.com/astropy/astropy/issues/7404 

1.0.0 (2021-04-04)
------------------
- Changed format of chain files, incompatible with previous versions.  Support for stars not in HGCA, ability to set one jitter per instrument.  Various small changes and bug fixes.  Renamed repo and orbit3d directory to orvara.

0.1.5 (2021-02-26)
------------------
- Added the ability to place mass priors on the secondary, tertiary, etc.. companions.

0.1.4 (2021-02-26)
------------------
- No changes, incorrect tag.

0.1.3 (2021-02-21)
------------------
- Added documentation.

0.1.2 (2021-01-28)
------------------
- Fixed bug where priors on the primary mass were applied to the primary mass + masses of interior companions.

0.1.1 (2019-08-15)
------------------
- Refactored code and added documentation. Bumped requirement
for HTOF to 0.1.1.

0.1.0 (prior to 2019-08-15)
---------------------------
- Initial release.


orvara/main.py
#!/usr/bin/env python
"""
Orbit fitting code. The run function is the console entry point,
accessed by calling fit_orbit from the command line.
"""

from __future__ import print_function
import numpy as np
import os
import time
import emcee
from ptemcee import Sampler as PTSampler
from configparser import ConfigParser
from astropy.io import fits
from htof.main import Astrometry
from astropy.time import Time
from astropy.coordinates import Angle
import sys
import re
import random
from orvara import orbit
from orvara.config import parse_args
from orvara.format_fits import make_header, pack_cols
import pkg_resources

_loglkwargs = {}

def set_initial_parameters(start_file, ntemps, nplanets, nwalkers, njit=1,
                           minjit=-20, maxjit=20):
    
    par0 = np.ones((ntemps, nwalkers, 2 + 7 * nplanets))

    if start_file.lower() == 'none':
        mpri = 1
        jit = 0.5
        sau = 10
        esino = 0.5
        ecoso = 0.5
        inc = 1
        asc = 1
        lam = 1
        msec = 0.1

        sig = np.ones((ntemps, nwalkers, 2 + 7 * nplanets))*0.5    
        init = [jit, mpri]
        for i in range(nplanets):
             init += [msec, sau, esino, ecoso, inc, asc, lam]
        par0 *= np.asarray(init)

    else:
        init, sig = np.loadtxt(start_file).T
        init[0] = max(2*np.log10(init[0]), minjit) # Convert from m/s to units used in code
        try:
            par0 *= init
        except:
            raise ValueError('Starting file %s has the wrong format/length.' % (start_file))

    #######################################################################
    # Introduce scatter.  Normal in most parameters, lognormal in
    # mass and semimajor axis.
    #######################################################################
    
    scatter = sig*np.random.randn(np.prod(par0.shape)).reshape(par0.shape)
    par0 += scatter
    par0[..., 2::7] = (par0[..., 2::7] - scatter[..., 2::7])*np.exp(scatter[..., 2::7])
    par0[..., 3::7] = (par0[..., 3::7] - scatter[..., 3::7])*np.exp(scatter[..., 3::7])

    # Ensure that values are within allowable ranges.
    
    bounds = [[0, minjit, maxjit],   # jitter
              [1, 1e-4, 1e3],        # mpri (Solar masses)
              [2, 1e-4, 1e3],        # msec (Solar masses)
              [3, 1e-5, 2e5],        # semimajor axis (AU)
              [6, 1e-5, np.pi],      # inclination (radians)
              [7, -np.pi, 3*np.pi],  # longitude of ascending node (rad)
              [8, -np.pi, 3*np.pi]]  # long at ref epoch (rad)
        
    for i in range(len(bounds)):
        j, minval, maxval = bounds[i]
        if j <= 1:
            par0[..., j][par0[..., j] < minval] = minval
            par0[..., j][par0[..., j] > maxval] = maxval
        else:
            par0[..., j::7][par0[..., j::7] < minval] = minval
            par0[..., j::7][par0[..., j::7] > maxval] = maxval
            
    # Eccentricity is a special case.  Cap at 0.99.
    ecc = par0[..., 4::7]**2 + par0[..., 5::7]**2
    fac = np.ones(ecc.shape)
    fac[ecc > 0.99] = np.sqrt(0.99)/np.sqrt(ecc[ecc > 0.99])
    par0[..., 4::7] *= fac
    par0[..., 5::7] *= fac

    # Move jitter to the end, add (shuffled) realizations if needed.
    par0_jitlast = np.zeros((ntemps, nwalkers, par0.shape[-1] + njit - 1))
    par0_jitlast[..., :-njit] = par0[..., 1:]
    for i in range(njit):
        random.shuffle(par0[..., 0].T)
        par0_jitlast[..., -1 - i] = par0[..., 0]

    return par0_jitlast


def initialize_data(config, companion_gaia):
    # load in items from the ConfigParser object
    HipID = config.getint('data_paths', 'HipID', fallback=0)
    HGCAFile = config.get('data_paths', 'HGCAFile')
    if not os.path.exists(HGCAFile):
        raise FileNotFoundError(f'No HGCA file found at {HGCAFile}')

    try:
        table = fits.open(HGCAFile)[1].data
        epoch_ra_gaia = table[table['hip_id'] == 1]['epoch_ra_gaia']
        if np.isclose(epoch_ra_gaia, 2015.60211565):
            HGCAVersion = 'GaiaDR2'
            gaia_mission_length_yrs = 1.75
        elif np.isclose(epoch_ra_gaia, 2015.92749023):
            HGCAVersion = 'GaiaeDR3'
            gaia_mission_length_yrs = 2.76
            if 'gaia_npar' in table.names:
                # TODO change this to GaiaDR3 when HTOF has a 'GaiaDR3' parser.
                #  right now for testing, gaiaedr3 works fine because the baseline of dr3 and edr3 will be the same.
                HGCAVersion = 'GaiaeDR3'
                gaia_mission_length_yrs = 2.76
        else:
            raise ValueError("Cannot match %s to either DR2 or eDR3, or DR3 based on RA epoch of Gaia" % (HGCAFile))
    except:
        raise ValueError("Cannot access HIP 1 in HGCA file" + HGCAFile)

    RVFile = config.get('data_paths', 'RVFile', fallback='')
    relRVFile = config.get('data_paths', 'relRVFile', fallback='')
    AstrometryFile = config.get('data_paths', 'AstrometryFile', fallback='')
    GaiaDataDir = config.get('data_paths', 'GaiaDataDir', fallback='')
    Hip2DataDir = config.get('data_paths', 'Hip2DataDir', fallback='')
    Hip1DataDir = config.get('data_paths', 'Hip1DataDir', fallback='')
    use_epoch_astrometry = config.getboolean('mcmc_settings', 'use_epoch_astrometry', fallback=False)
    data = orbit.Data(HipID, HGCAFile, RVFile, AstrometryFile, relRVFile=relRVFile, companion_gaia=companion_gaia,
                      gaia_mission_length_yrs=gaia_mission_length_yrs)
    if use_epoch_astrometry and data.use_abs_ast == 1:
        # five-parameter fit means a first order polynomial, 7-parameter means 2nd order polynomial etc..
        gaia_fit_degree = {5: 1, 7: 2, 9: 3}[data.gaia_npar]
        Gaia_fitter = Astrometry(HGCAVersion, '%06d' % (HipID), GaiaDataDir,
                                 central_epoch_ra=data.epRA_G,
                                 central_epoch_dec=data.epDec_G,
                                 format='jyear', fit_degree=gaia_fit_degree,
                                 use_parallax=True, use_catalog_parallax_factors=True)
        Hip2_fitter = Astrometry('Hip2or21', '%06d' % (HipID), Hip2DataDir,
                                 central_epoch_ra=data.epRA_H,
                                 central_epoch_dec=data.epDec_H,
                                 format='jyear', use_parallax=True, use_catalog_parallax_factors=True)
        Hip1_fitter = Astrometry('Hip1', '%06d' % (HipID), Hip1DataDir,
                                 central_epoch_ra=data.epRA_H,
                                 central_epoch_dec=data.epDec_H,
                                 format='jyear', use_parallax=True, use_catalog_parallax_factors=True)
        # instantiate C versions of the astrometric fitter which are much faster than HTOF's Astrometry
        hip1_fast_fitter = orbit.AstrometricFitter(Hip1_fitter)
        hip2_fast_fitter = orbit.AstrometricFitter(Hip2_fitter)
        gaia_fast_fitter = orbit.AstrometricFitter(Gaia_fitter)
        # generate the data object.
        data = orbit.Data(HipID, HGCAFile, RVFile, AstrometryFile, 
                          use_epoch_astrometry=use_epoch_astrometry,
                          relRVFile=relRVFile,
                          epochs_Hip1=Hip1_fitter.data.julian_day_epoch(),
                          epochs_Hip2=Hip2_fitter.data.julian_day_epoch(),
                          epochs_Gaia=Gaia_fitter.data.julian_day_epoch(),
                          companion_gaia=companion_gaia, verbose=False)
    elif data.use_abs_ast == 1:
        hip1_fast_fitter, hip2_fast_fitter, gaia_fast_fitter = None, None, None
    else:
        hip1_fast_fitter, hip2_fast_fitter, gaia_fast_fitter = None, None, None
        try:
            data.plx = 1e-3*config.getfloat('priors_settings', 'parallax')
            data.plx_err = 1e-3*config.getfloat('priors_settings', 'parallax_error')
        except:
            raise RuntimeError("Cannot load absolute astrometry. Please supply a prior "
                               "for parallax and parallax_error in the priors_settings area"
                               " of the configuration file.")
    # If at any point, we implement jerk fitting to the calc_pms_no_epoch_astrometry(), this catch can be removed:
    if data.gaia_npar == 9 and not use_epoch_astrometry:
        raise RuntimeError("This is a 9-parameter source in Gaia, but you have set use_epoch_astrometry=False"
                           " in the configuration file. Please enable use_epoch_astrometry=True for this source, and "
                           "follow the directions in section 'Epoch Astrometry' of the readme.")
    return data, hip1_fast_fitter, hip2_fast_fitter, gaia_fast_fitter


def lnprob(theta, returninfo=False, RVoffsets=False, use_epoch_astrometry=False,
           data=None, nplanets=1, H1f=None, H2f=None, Gf=None, priors=None, 
           njitters=1):
    """
    Log likelihood function for the joint parameters
    :param theta:
    :param returninfo:
    :param use_epoch_astrometry:
    :param data:
    :param nplanets:
    :param H1f:
    :param H2f:
    :param Gf:
    :return:
    """
    model = orbit.Model(data)
    lnp = 0
    for i in range(nplanets):
        # Note that params.mpri is really the mass contained in the primary + companions inner to the current planet.
        # params.mpri_true is the real mass of the primary. So params.mpri should really be renamed params.interior_mass
        params = orbit.Params(theta, i, nplanets, data.nInst, njitters)
        lnp = lnp + orbit.lnprior(params, minjit=priors['minjit'],
                                  maxjit=priors['maxjit'])

        if not np.isfinite(lnp):
            model.free()
            params.free()
            return -np.inf

        orbit.calc_EA_RPP(data, params, model)
        orbit.calc_RV(data, params, model)
        if data.n_rel_RV > 0:
            # slightly more performant if we ignore the relative RV calculation call entirely if
            # we do not have any relative RVs to calculate.
            orbit.calc_relRV(data, params, model, i)
        orbit.calc_offsets(data, params, model, i)

        chisq_sec = (params.msec - priors[f'm_secondary{i}'])**2/priors[f'm_secondary{i}_sig']**2
        if chisq_sec > 0:  # If chisq_sec > 0 then a prior was set on the mass of the secondary
            # undo the default 1/m prior that is set in orbit.pyx (see lnprior())
            lnp = lnp - 0.5*chisq_sec + np.log(params.msec)
        # else, don't change lnp at all.

        # Free params if we need to cycle through the next companion
        if i < nplanets - 1:
            params.free()

    if use_epoch_astrometry and data.use_abs_ast:
        orbit.calc_PMs_epoch_astrometry(data, model, H1f, H2f, Gf)
    elif data.use_abs_ast:
        orbit.calc_PMs_no_epoch_astrometry(data, model)

    if returninfo:
        return orbit.calcL(data, params, model, chisq_resids=True, RVoffsets=RVoffsets)

    if np.isfinite(priors['mpri_sig']):
        return lnp - 0.5*(params.mpri_true - priors['mpri'])**2/priors['mpri_sig']**2 + orbit.calcL(data, params, model)
    else:
        return lnp - np.log(params.mpri_true) + orbit.calcL(data, params, model)

    
def return_one(theta):
    return 1.


def avoid_pickle_lnprob(theta):
    global _loglkwargs
    return lnprob(theta, **_loglkwargs)


def get_priors(config):
    priors = {}
    priors['mpri'] = config.getfloat('priors_settings', 'mpri', fallback=1.)
    priors['mpri_sig'] = config.getfloat('priors_settings', 'mpri_sig', fallback=np.inf)
    # priors on the masses of the companions (labelled 0 though 9). Limit to 10 planet systems.
    for i in range(10):
        priors[f'm_secondary{i}'] = config.getfloat('priors_settings', f'm_secondary{i}', fallback=1.)
        priors[f'm_secondary{i}_sig'] = config.getfloat('priors_settings', f'm_secondary{i}_sig', fallback=np.inf)
    # priors on the RV jitter. Converting from m/s to orvara internal units.
    priors['minjit'] = config.getfloat('priors_settings', 'minjitter', fallback = 1e-5)
    priors['minjit'] = max(priors['minjit'], 1e-20) # effectively zero, but we need the log
    priors['minjit'] = 2*np.log10(priors['minjit'])
    priors['maxjit'] = config.getfloat('priors_settings', 'maxjitter', fallback = 1e3)
    priors['maxjit'] = 2*np.log10(priors['maxjit'])
    if priors['maxjit'] < priors['minjit']:
        raise ValueError("Requested maximum jitter < minimum jitter")
    return priors


def get_gaia_catalog_companion(config):
    companion_gaia = {}
    companion_gaia['ID'] = config.getint('secondary_gaia', 'companion_ID', fallback=-1)
    companion_gaia['pmra'] = config.getfloat('secondary_gaia', 'pmra', fallback=0)
    companion_gaia['pmdec'] = config.getfloat('secondary_gaia', 'pmdec', fallback=0)
    companion_gaia['e_pmra'] = config.getfloat('secondary_gaia', 'epmra', fallback=1)
    companion_gaia['e_pmdec'] = config.getfloat('secondary_gaia', 'epmdec', fallback=1)
    companion_gaia['corr_pmra_pmdec'] = config.getfloat('secondary_gaia', 'corr_pmra_pmdec', fallback=0)
    return companion_gaia


def run():
    """
    Initialize and run the MCMC sampler.
    """
    global _loglkwargs
    start_time = time.time()

    args = parse_args()
    config = ConfigParser()
    if not os.path.exists(args.config_file):
        raise FileNotFoundError(f'No config file found at {args.config_file}')
    if not os.path.exists(args.output_dir):
        raise FileNotFoundError(f'No output_dir found at {args.output_dir}')
    config.read(args.config_file)

    # set the mcmc parameters
    nwalkers = config.getint('mcmc_settings', 'nwalkers', fallback=100)
    ntemps = config.getint('mcmc_settings', 'ntemps', fallback=10)
    nplanets = config.getint('mcmc_settings', 'nplanets')
    jit_per_inst = config.getboolean('mcmc_settings', 'jit_per_inst', fallback=False)
    nstep = config.getint('mcmc_settings', 'nstep')
    thin = config.getint('mcmc_settings', 'thin', fallback=50)
    nthreads = config.getint('mcmc_settings', 'nthreads', fallback=1)
    use_epoch_astrometry = config.getboolean('mcmc_settings', 'use_epoch_astrometry', fallback=False)
    HipID = config.getint('data_paths', 'HipID', fallback=0)
    start_file = config.get('data_paths', 'start_file', fallback='none')
    starname = config.get('plotting', 'target', fallback='HIP%d' % (HipID))
    #
    priors = get_priors(config)
    companion_gaia = get_gaia_catalog_companion(config)
    
    # Save configuration file in FITS header format
    header = make_header(args.config_file)

    data, H1f, H2f, Gf = initialize_data(config, companion_gaia)

    # set initial conditions
    if jit_per_inst:
        njit = data.nInst
    else:
        njit = 1
    par0 = set_initial_parameters(start_file, ntemps, nplanets, nwalkers, 
                                  njit=njit, minjit=priors['minjit'], 
                                  maxjit=priors['maxjit'])
    ndim = par0[0, 0, :].size

    # set arguments for emcee PTSampler and the log-likelyhood (lnprob)
    samplekwargs = {'thin': thin}
    loglkwargs = {'returninfo': False, 'use_epoch_astrometry': use_epoch_astrometry,
                  'data': data, 'nplanets': nplanets, 'H1f': H1f, 'H2f': H2f, 'Gf': Gf, 'priors': priors, 'njitters': njit}
    _loglkwargs = loglkwargs
    # run sampler without feeding it loglkwargs directly, since loglkwargs contains non-picklable C objects.

    try:
        use_ptemcee = False
        sample0 = emcee.PTSampler(ntemps, nwalkers, ndim, avoid_pickle_lnprob, return_one, threads=nthreads)
        print('Using emcee.PTSampler.')
    except:
        use_ptemcee = True
        sample0 = PTSampler(ntemps=ntemps, nwalkers=nwalkers, dim=ndim,
                            logl=avoid_pickle_lnprob, logp=return_one,
                            threads=nthreads)
        print('Using ptemcee.')

    print("Running MCMC ... ")
    #sample0.run_mcmc(par0, nstep, **samplekwargs)
    #add a progress bar
    width = 30
    N = min(100, nstep//thin)
    n_taken = 0
    sys.stdout.write("[{0}]  {1}%".format(' ' * width, 0))
    for ipct in range(N):
        dn = (((nstep*(ipct + 1))//N - n_taken)//thin)*thin
        n_taken += dn
        if ipct == 0:
            sample0.run_mcmc(par0, dn, **samplekwargs)
        else:
            # Continue from last step
            sample0.run_mcmc(sample0.chain[..., -1, :], dn, **samplekwargs)
        n = int((width+1) * float(ipct + 1) / N)
        sys.stdout.write("\r[{0}{1}]".format('#' * n, ' ' * (width - n)))
        sys.stdout.write("%3d%%" % (int(100*(ipct + 1)/N)))
    sys.stdout.write("\n")
        
    print('Total Time: %.0f seconds' % (time.time() - start_time))
    print("Mean acceptance fraction (cold chain): {0:.6f}".format(np.mean(sample0.acceptance_fraction[0, :])))
    # save data
    if not use_ptemcee:
        shape = sample0.lnprobability[0].shape
    else:
        shape = sample0.logprobability[0].shape
    parfit = np.zeros((shape[0], shape[1], 9 + data.nInst))

    loglkwargs['returninfo'] = True
    loglkwargs['RVoffsets'] = True
    
    for i in range(shape[0]):
        for j in range(shape[1]):
            res, RVoffsets = lnprob(sample0.chain[0][i, j], **loglkwargs)
            parfit[i, j, :9] = [res.plx_best, res.pmra_best, res.pmdec_best,
                                res.chisq_sep, res.chisq_PA,
                                res.chisq_H, res.chisq_HG, res.chisq_G, res.chisq_relRV]
            
            if data.nInst > 0:
                parfit[i, j, 9:] = RVoffsets

    colnames = ['mpri']
    units = ['msun']
    for i in range(nplanets):
        colnames += [s + '%d' % (i) for s in ['msec', 'sau', 'esino', 'ecoso',
                                              'inc', 'asc', 'lam']]
        units += ['msun', 'au', '', '', 'radians', 'radians', 'radians']

    if njit == 1:
        colnames += ['jitter']
        units += ['m/s']
        sample0.chain[0][..., -1] = 10**(0.5*sample0.chain[0][..., -1])
    else:
        for i in range(njit):
            colnames += ['jitter%d' % (i)]
            units += ['m/s']
        sample0.chain[0][..., -njit:] = 10**(0.5*sample0.chain[0][..., -njit:])

    colnames += ['lnp']
    colnames += ['plx_ML', 'pmra_ML', 'pmdec_ML', 'chisq_sep', 
                 'chisq_PA', 'chisq_H', 'chisq_HG', 'chisq_G', 'chisq_relRV']
    units += ['', 'arcsec', 'arcsec/yr', 'arcsec/yr', '', '', '', '', '', '']
    colnames += ['RV_ZP_%d_ML' % (i) for i in range(data.nInst)]
    units += ['m/s' for i in range(data.nInst)]

    out = fits.HDUList(fits.PrimaryHDU(None, header))

    if not use_ptemcee:
        lnp = sample0.lnprobability[0]
    else:
        lnp = sample0.logprobability[0]

    out.append(pack_cols(sample0.chain[0], lnp, parfit, colnames, units))

    for i in range(1000):
        filename = os.path.join(args.output_dir, '%s_chain%03d.fits' % (starname, i))
        if not os.path.isfile(filename):
            print('Writing output to {0}'.format(filename))
            out.writeto(filename, overwrite=False)
            break

    return out


if __name__ == "__main__":
    run()
